atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -326,26 +325,23 @@ public void registerClosingCallback(Runnable closingCallback) {
      */
     @Override
     public void close() {
-        if (isClosed.compareAndSet(false, true)) {
-            super.close();
-            cleaner.close();
-            keyValueService.close();
-            shutdownExecutor(deleteExecutor);
-            shutdownExecutor(getRangesExecutor);
-            closeLockServiceIfPossible();
-
-            List<Throwable> suppressedExceptions = new ArrayList<>();
+        if (!isClosed.compareAndSet(false, true)) {
+            return;
+        }
+
+        try (ShutdownRunner shutdownRunner = new ShutdownRunner()) {
+            shutdownRunner.shutdownSafely(super::close);
+            shutdownRunner.shutdownSafely(cleaner::close);
+            shutdownRunner.shutdownSafely(keyValueService::close);
+            shutdownRunner.shutdownSafely(() -> shutdownExecutor(deleteExecutor));
+            shutdownRunner.shutdownSafely(() -> shutdownExecutor(getRangesExecutor));
+            shutdownRunner.shutdownSafely(this::closeLockServiceIfPossible);
+
             for (Runnable callback : Lists.reverse(closingCallbacks)) {
-                runShutdownCallbackSafely(callback).ifPresent(suppressedExceptions::add);
+                shutdownRunner.shutdownSafely(callback);
```

Comment:
Given that there is a test for this class, would you be able to add any tests that verify this new behaviour (i.e. what happens when an exception is thrown during close)?

Reply:
I will see what tests I can add, although there's already a whole bunch of tests for SnapshotTransactionManager to do with #close behavior

Reply:
If I am not mistaken, none of them handle errors during the close.

Reply:
Tests added, good shout.